### PR TITLE
Cosine Similarity UDF function

### PR DESCRIPTION
**Pull request by daniellevy**

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
 
         return lower;
     }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double cosineSimilarity(@Nullable @SqlType(MAP_STRING_DOUBLE) Block map1, @Nullable @SqlType(MAP_STRING_DOUBLE) Block map2)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: These should not be nullable

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+    {
+        Map<String, Double> m = toStringDoubleMap(map1);
+        Map<String, Double> n = toStringDoubleMap(map2);
+
+        double norm_m = Math.sqrt(dotProduct(m, m));
+        double norm_n = Math.sqrt(dotProduct(n, n));
+
+        if(norm_m == 0 || norm_n == 0)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Formatting

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+            return 0;
+        }
+
+        double dotProduct = dotProduct(m, n);
+
+        return dotProduct/(norm_m * norm_n);
+
+    }
+
+    public static Map<String, Double> toStringDoubleMap(Block map)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Private

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 4]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+    {
+        Map<String, Double> hashmap = new HashMap<>();
+
+        if (map != null) {
+            for (int position = 0; position < map.getPositionCount(); position += 2)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: This won't be null

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 5]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
 
         return lower;
     }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double cosineSimilarity(@Nullable @SqlType(MAP_STRING_DOUBLE) Block map1, @Nullable @SqlType(MAP_STRING_DOUBLE) Block map2)
+    {
+        Map<String, Double> m = toStringDoubleMap(map1);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Why strings? These should stay as doubles

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 6]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+        }
+
+        return hashmap;
+    }
+
+    public static double dotProduct(Map<String, Double> m, Map<String, Double> n)
+    {
+        double result = 0;
+
+        if(m == null || n == null)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Remove, won't happen

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 7]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+
+    public static Map<String, Double> toStringDoubleMap(Block map)
+    {
+        Map<String, Double> hashmap = new HashMap<>();
+
+        if (map != null) {
+            for (int position = 0; position < map.getPositionCount(); position += 2)
+            {
+                Slice slice = VARCHAR.getSlice(map, position);
+                String key = slice.toStringUtf8();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Use Slice as the key

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 8]
```diff
@@ -692,4 +699,63 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
+
+        double dotProduct = dotProduct(m, n);
+
+        return dotProduct/(norm_m * norm_n);
+
+    }
+
+    public static Map<String, Double> toStringDoubleMap(Block map)
+    {
+        Map<String, Double> hashmap = new HashMap<>();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: See if there is a more efficient map type in fastutils

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 9]
```diff
@@ -772,4 +772,24 @@ public void testWidthBucketArray()
         // this is a case that we can't catch because we are using binary search to bisect the bins array
         assertFunction("width_bucket(1.5, array[1.0, 2.3, 2.0])", BIGINT, 1L);
     }
+
+    @Test
+    public void testCosineSimilarity()
+    {
+//        assertFunction("cosine_similarity( map(array ['a', 'b'], array [1.0, 2.0]), map(array ['c', 'b'], array [1.0, 3.0]) )",
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: These look useful. Why commented out?

We don't add commented out code. Either remove or uncomment

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 10]
```diff
@@ -772,4 +772,24 @@ public void testWidthBucketArray()
...
+//                DOUBLE,
+//                2 * 3 / (Math.sqrt(5) * Math.sqrt(10) ) );
+//
+//        assertFunction("cosine_similarity( map(array ['a', 'b', 'c'], array [1.0, 2.0, -1.0]), map(array ['c', 'b'], array [1.0, 3.0]) )",
+//                DOUBLE,
+//                (2 * 3 + (-1) * 1 ) / (Math.sqrt(1+4+1) * Math.sqrt(1 + 9) ) );
+
+        assertFunction("cosine_similarity( null, map(array ['c', 'b'], array [1.0, 3.0]) )",
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: This function needs more than one test

File: presto-ml/src/main/java/com/facebook/presto/ml/MLFunctions.java [hunk 11]
```diff
@@ -31,6 +31,8 @@
 import io.airlift.slice.Slice;
 import io.airlift.slice.Slices;
 
+import java.util.Map;
```
Comment (electrum) [inline presto-ml/src/main/java/com/facebook/presto/ml/MLFunctions.java]: These ML changes are unrelated and should be removed

**Comment by electrum** [comment]
Please add this new function to the documentation

**Comment by raghavsethi** [comment]
@electrum I'm working with @daniellevy to get this in.

**Comment by raghavsethi** [comment]
Please squash your commits

**Comment by raghavsethi** [comment]
Also make sure the Travis run is clean
